frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -649,6 +649,144 @@ const ResolvedFields& fieldsForTypeDecl(Context* context,
   return f;
 }
 
+static const owned<ResolvedForwarding>&
+forwardingForTypeDeclQuery(Context* context, const CompositeType* ct) {
+  QUERY_BEGIN(forwardingForTypeDeclQuery, context, ct);
+
+  owned<ResolvedForwarding> result;
+
+  const AggregateDecl* ad = nullptr;
+  if (auto ast = parsing::idToAst(context, ct->id())) {
+    ad = ast->toAggregateDecl();
+  }
+
+  if (ad != nullptr) {
+    ResolutionResultByPostorderID rr;
+    auto visitor = Resolver::createForForwarding(context, ad, ct, rr);
+    // Performance: do the parent type & fields really need to be visited here?
+
+    // visit the parent type
+    if (auto cls = ad->toClass()) {
+      if (auto parentClassExpr = cls->parentClass()) {
+        parentClassExpr->traverse(visitor);
+      }
+    }
+    // visit the field declarations & forwarding declarations
+    for (auto child: ad->children()) {
+      if (child->isVariable() ||
+          child->isMultiDecl() ||
+          child->isTupleDecl() ||
+          child->isForwardingDecl()) {
+        child->traverse(visitor);
+      }
+    }
```

Comment:
All of this reminds me a lot of the `fieldsForTypeDecl` query - could we use it here, and then simply search the field IDs for the `fwd->expr()`'s ID? We might be able to share code (and cached computation results) this way.

Reply:
Well, I agree with you that it might be worthwhile to run the resolver once to resolve the fields instead of twice. The reason that I did not do that is that it would imply more coupling between the tasks than is strictly necessary, and I was worried that would cause problems with the recursion detection for forwarding. Additionally, I am expecting that forwarding is relatively rarely used, so having a more robust implementation is better than saving some time.

But, I'll respond to your suggestion here by adding a TODO comment here.